packages/shell/src/views/DebuggerView.ts

```diff
@@ -0,0 +1,1191 @@
+    } else {
+      newSet.add(index);
+    }
+    this.expandedEvents = newSet;
+  }
+
+  private toggleJsonFullHeight(index: number) {
+    const newSet = new Set(this.fullHeightEvents);
+    if (newSet.has(index)) {
+      newSet.delete(index);
+    } else {
+      newSet.add(index);
+    }
+    this.fullHeightEvents = newSet;
+  }
+
+  private async copyJson(data: any) {
+    try {
+      const jsonString = JSON.stringify(data, null, 2);
+      await navigator.clipboard.writeText(jsonString);
+    } catch (err) {
+      console.error("Failed to copy JSON:", err);
+    }
+  }
+
+  private formatTime(timestamp: number): string {
+    const date = new Date(timestamp);
+    return `${date.toLocaleTimeString()}.${
+      date.getMilliseconds().toString().padStart(3, "0")
+    }`;
+  }
+
+  private getEventIcon(marker: RuntimeTelemetryMarkerResult): string {
+    const type = marker.type;
+
+    // Try to find a matching topic
+    for (const [topicKey, topic] of Object.entries(TOPIC_HIERARCHY)) {
+      for (const [subtopicKey, subtopic] of Object.entries(topic.subtopics)) {
+        if (type.startsWith(subtopic.pattern)) {
+          return topic.icon;
+        }
+      }
+    }
+
+    // Default icon
+    return "📊";
+  }
+
+  private getEventColor(marker: RuntimeTelemetryMarkerResult): string {
+    const type = marker.type;
+
+    // Try to find a matching topic
+    for (const [topicKey, topic] of Object.entries(TOPIC_HIERARCHY)) {
+      for (const [subtopicKey, subtopic] of Object.entries(topic.subtopics)) {
+        if (type.startsWith(subtopic.pattern)) {
+          return topic.color;
+        }
+      }
+    }
+
+    // Default color
+    return "#64748b";
+  }
+
+  private matchesActiveTopics(marker: RuntimeTelemetryMarkerResult): boolean {
+    if (this.activeSubtopics.size === 0) return false;
+
+    const type = marker.type;
+
+    // Check if the event matches any active subtopic
+    for (const [topicKey, topic] of Object.entries(TOPIC_HIERARCHY)) {
+      for (const [subtopicKey, subtopic] of Object.entries(topic.subtopics)) {
+        const fullKey = `${topicKey}.${subtopicKey}`;
+        if (
+          this.activeSubtopics.has(fullKey) && type.startsWith(subtopic.pattern)
+        ) {
+          return true;
+        }
+      }
+    }
+
+    return false;
+  }
+
+  private matchesSearch(marker: RuntimeTelemetryMarkerResult): boolean {
+    if (!this.searchText) return true;
+
+    const searchLower = this.searchText.toLowerCase();
+    const markerStr = JSON.stringify(marker).toLowerCase();
+    return markerStr.includes(searchLower);
+  }
+
+  private getFilteredEvents(): RuntimeTelemetryMarkerResult[] {
+    const markers = this.isPaused ? this.pausedMarkers : this.telemetryMarkers;
+
+    return markers.filter((marker) =>
+      this.matchesActiveTopics(marker) && this.matchesSearch(marker)
+    );
+  }
+
+  private renderEventDetails(marker: RuntimeTelemetryMarkerResult): any {
+    const details = [];
```

Comment:
Can imagine this function being the guts of a `DebuggerMarkerView` in the future, handling rendering of different marker types